include/mbedtls/ecp.h

```diff
@@ -501,6 +501,50 @@ int mbedtls_ecp_tls_read_group( mbedtls_ecp_group *grp, const unsigned char **bu
 int mbedtls_ecp_tls_write_group( const mbedtls_ecp_group *grp, size_t *olen,
                          unsigned char *buf, size_t blen );
 
+#if defined(MBEDTLS_ASN1_WRITE_C) && defined(MBEDTLS_OID_C)
+/**
+ * \brief           Maximum size of the output of mbedtls_ecp_ansi_write_group
+ */
+#define MBEDTLS_OID_EC_GRP_MAX_SIZE 12
```

Comment:
Symbols starting with `MBEDTLS_OID_` should be defined in the OID, not ECP, module. Also, the value `12` is entirely magical and unexplained. Perhaps moving this definition in the OID module next to the definition of the actual OID would help (and adding a comment about the +2).

Reply:
Which place exactly do you mean from oid.h?

Reply:
From the comment, it looks like `12` is 2 + maximum of `sizeof` of some OIDs. By reading the implementation of `mbedtls_ecp_ansi_write_group()` and then of `mbedtls_oid_get_oid_by_ec_grp` I gather the OIDs in question are the various `MBEDTLS_OID_EC_GRP_xxx` defined between line 289 and 334 of `oid.h`. I think a good way to go would be to define `MBEDTLS_OID_EC_GRP_MAX_SIZE` to that maximum right after line 334 (and add a comment on line 284 reminding people to update it when adding new curves). Then in `ecp.h` we could define `MBEDTLS_ECP_GRP_OID_MAX_SIZE` to the previous one plus 2 (with a comment explaining why `2`).

Actually I'm wondering if hardcoding a maximum computed by hand is a good idea or if we should preprocessor tricks or a function that iterates over `oid_ecp_grp` to compute it. It would certainly be safer but I'm not sure if it's worth the trouble. @gilles-peskine-arm what do you think?

Reply:
Sorry for not leaving a comment explaining the value. The value is the maximum size of the OID of a supported group, plus 2 for tag and length.

I did consider computing this, but it looked very complicated: lots of constants to grab, each under its own conditional preprocessing directive. And it would give better documentation, but not better safety, because the biggest risk here is to forget one.

Reply:
I do understand where the value comes from, my concern is that it's not defined close to where the value comes from, which IMO increases the risk of messing things up if and when we add a new curve.

(That plus the name which doesn't match the file where it's defined.)

Reply:
Manuel - your solution will introduce an additional dependency - ecp.h will depend on oid.h (due to the MBEDTLS_OID_EC_GRP_MAX_SIZE definition). Is it intended?

Reply:
Unfortunately this creates a circular dependency:
ecdh.h -> ecp.h -> oid.h -> pk.h -> ecp.h
Even if oid.h wouldn't depend on pk.h - oid.h still depends on x509.h, and it depends on pk.h.
How about leaving the define "#define MBEDTLS_ECP_GRP_OID_MAX_SIZE 12" in ecp.h, documenting it, and also writing a comment (the one you've mentioned would be useful for people adding a new curve) in oid.h, so that it points to this define?

Reply:
I agree with @AndrzejKurek's proposal: keep the constant in `ecp.h`, with an `MBEDTLS_ECP_` name.

The fact that this constant involves the size of an OID is actually an implementation detail. What matters is that it's a bound on the size of the output of `mbedtls_ecp_ansi_write_group`. In theory, this function could use other formats (in practice, this would break interoperability with other parts of the library).

Note that the constant needs to be _at least_ large enough for all the curves that the library supports, but we may want to make it larger, so in case people declare custom curves with larger OIDs, or so that we can add new curves with larger OIDs in a future library version without breaking the ABI.

Is there a comprehensive register of elliptic curve OIDs somewhere? If not https://www.flexiprovider.de/CurveOIDs.html looks like a good start.

Reply:
I'm OK with keeping the constant it `ecp.h` (but then it need to have an ECP name), at the condition that we add a clear and visible comment in the relevant section in `oid.h` - I really don't like this kind of non-locality where editing one file means you have to remember to edit a constant in another file, so I think it really needs to be clearly documented.

I don't really see a need to care for people "declaring custom curves" as that's not something we support, but I agree that we might want to allow some room for adding curves in the future without breaking the ABI. Note that the above reference misses the curves we're IMO most likely to add in the future, which are here: https://tools.ietf.org/html/draft-ietf-curdle-pkix-06#section-3 (not final yet). Fortunately the provisional OIDs there are short.

Reply:
Updated to the value of 30 with a relevant documentation.